### ISSUE TITLE
something wrong with performance

### ISSUE THREAD

**Issue by datistiquo** [UKPLab/sentence-transformers#606]
Hey,

that is weird. I try the bi encoder model training for my own small data set with just few thousand of lines.

But adapting this example:
https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/ms_marco/train_bi-encoder.py

and running on cholab with GPU is too slow. Seems it does not use the gpu?

What is also weird is the number of maximum iterations which is totally wrong:

after few minutes I just have this number of iterations:

152/6206657 [00:34<391:47:26, 4.40it/s]

And this number 6206657  is wrong when it is the number of batches. This should be below 100 in my case...

I come from tensorflow and maybe I do oversee something?  I do not know what is wrong here as I assumed most of this is self explainable.

**Comment by nreimers** [comment]
This
https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/ms_marco/train_bi-encoder.py#L160

defines the size of the dataset.

As we want to avoid reading the full MS MARCO file, I hard coded the size there. If you have a smaller dataset, you have to change that respective line.

**Comment by datistiquo** [comment]
> defines the size of the dataset.

Oh ok. :-)

I feel you have to have there many insight knowledge. So, this is the first example how to use the MultipleNegativeRankingLoss with triplets? And, if using triplets the negative one should be hard negatives.?
When having no hard negative you have to rely using the standard iterator with just pairs of (query, positive), right?
Because I applied this example by explicitly using triplets created extra for this but with normal random negatives as not hard negatives but just random negatives... 

If using the triplet form does the loss just  take all negatives in one pot (random+hard ones)?

Also, I find this lines very strange as it looks like all have the label "1"

https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/ms_marco/train_bi-encoder.py#L149

label = torch.tensor(1, dtype=torch.long)
But what is now the meaning of this lneght? How do apply this custom triplet data iterator to my case?

**Comment by nreimers** [comment]
You can also use it only with positive pairs:
(query, answer)

Adding a random negative entry would there not be advisable.  It only makes sense to use triplets if you have hard negatives.

With pairs, it will use all other entry as negatives.

The label is ignored, any value could be used.

**Comment by datistiquo** [comment]
So here I could also omit the label?

```
######### Read train data  ##########
train_samples = []
with open(os.path.join(dataset_path, "classification/train_pairs.tsv"), encoding='utf8') as fIn:
    reader = csv.DictReader(fIn, delimiter='\t', quoting=csv.QUOTE_NONE)
    for row in reader:
        if row['is_duplicate'] == '1':
            train_samples.append(InputExample(texts=[row['question1'], row['question2']], label=1))
            train_samples.append(InputExample(texts=[row['question2'], row['question1']], label=1)) #if A is a duplicate of B, then B is a duplicate of A


# After reading the train_samples, we create a SentencesDataset and a DataLoader
train_dataset = SentencesDataset(train_samples, model=model)
train_dataloader = DataLoader(train_dataset, shuffle=True, batch_size=train_batch_size)
train_loss = losses.MultipleNegativesRankingLoss(model)
```

**Comment by nreimers** [comment]
Some older version required that a label is passed to InputExample. But the value of the label is ignored.

The length are the number of pairs / triplets you have in total.

**Comment by datistiquo** [comment]
So it is the number before the loss generates some new data.

So with eg 100 pairs and batch size of 64 I get 4 batches for each epoch. How does this work out? And I suppose the 63 negative examples for each bacth are drawn randomly. But if this yields 4 batches I just trained 4 positive examples for each epoch out of 100?!

**Comment by nreimers** [comment]
The loss does not generate new data. It uses the other examples in a batch as negative as described in the linked paper.

In each batch, for each example, you have one positive and 63 negatives. So in total in a batch you have 64 positives and 64*63 negatives.

**Comment by datistiquo** [comment]
So for total 100 (positive) examples I would guess to have just 2 batches and not 4?

**Comment by nreimers** [comment]
yes

**Comment by datistiquo** [comment]
Yes, just saw that in your quora example for this loss you put also B,A for A,b as example. That is why the bacthes is doubled. :-)
